text/0000-non-exhaustive.md

```diff
@@ -0,0 +1,548 @@
+
+match error {
+    Message(ref s) => ...,
+    Other => ...,
+    _ => ...,
+}
+```
+
+And it should *not* be marked as dead code, even if the compiler does mark it as
+dead and remove it.
+
+Note that this can *potentially* cause breaking changes if a user adds
+`#[deny(dead_code)]` to a match statement *and* the upstream crate removes the
+`#[non_exhaustive]` lint. That said, modifying warn-only lints is generally
+assumed to not be a breaking change, even though users can make it a breaking
+change by manually denying lints.
+
+## Structs
+
+Like with enums, the attribute is essentially ignored in the crate that defines
+the struct, so that users can continue to construct values for the struct.
+However, this will prevent downstream users from constructing or exhaustively
+matching the struct, because fields may be added to the struct in the future.
+
+Additionally, adding `#[non_exhaustive]` to an enum variant will operate exactly
+the same as if the variant were a struct.
+
+Using our `Config` again:
+
+```rust
+#[non_exhaustive]
+pub struct Config {
+    pub window_width: u16,
+    pub window_height: u16,
+}
+```
+
+We can still construct our config within the defining crate like so:
+
+```rust
+let config = Config { window_width: 640, window_height: 480 };
+```
+
+And we can even exhaustively match on it, like so:
+
+```rust
+if let Ok(Config { window_width, window_height }) = load_config() {
+    // ...
+}
+```
+
+But users outside the crate won't be able to construct their own values, because
+otherwise, adding extra fields would be a breaking change.
+
+Users can still match on `Config`s non-exhaustively, as usual:
+
+```rust
+let &Config { window_width, window_height, .. } = config;
+```
+
+But without the `..`, this code will fail to compile.
+
+Although it should not be explicitly forbidden by the language to mark a struct
+with some private fields as non-exhaustive, it should emit a warning to tell the
+user that the attribute has no effect.
+
+## Tuple structs
+
+Non-exhaustive tuple structs will operate similarly to structs, however, will
+disallow matching directly. For example, take this example on stable today:
+
+```rust
+pub Config(pub u16, pub u16, ());
+```
+
+The below code does not work, because you can't match tuple structs with private
+fields:
+
+```rust
+let Config(width, height, ..) = config;
+```
+
+However, this code *does* work:
+
+```rust
+let Config { 0: width, 1: height, .. } = config;
+```
+
+So, if we label a struct non-exhaustive:
+
+```
+#[non_exhaustive]
+pub Config(pub u16, pub u16)
+```
+
+Then we the only valid way of matching will be:
+
+```rust
+let Config { 0: width, 1: height, .. } = config;
+```
+
+## Unit structs
+
+Unit structs will work very similarly to tuple structs. Consider this struct:
+
+```rust
+#[non_exhaustive]
+pub struct Unit;
+```
+
+We won't be able to construct any values of this struct, but we will be able to
+match it like:
+
+```rust
+let Unit { .. } = unit;
```

Comment:
Same as previously, the constructor's visibility need to be adjusted and FRU is usable.